File: packages/transport/mcp/python/caracalai_transport_mcp/authenticate.py
# Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
# Caracal, a product of Garudex Labs
#
# Transport-neutral MCP authentication: identity verify and revocation check.

from __future__ import annotations

import re

from caracalai_identity import (
    AgentIdentityRequiredError,
    ChainMismatchError,
    DelegationRequiredError,
    HopCountExceededError,
    JwtConfig,
    ScopeInsufficientError,
    TokenInvalidError,
    ZoneInvalidError,
    verify_config,
)
from caracalai_revocation import RevocationStore

from .types import AuthError, AuthResult


def extract_bearer(auth_header: str | None) -> str | None:
    if not auth_header:
        return None
    match = re.match(r"^Bearer\s+(.+)$", auth_header, re.IGNORECASE)
    if not match:
        return None
    token = match.group(1).strip()
    return token or None


async def authenticate(
    token: str,
    issuer: str,
    audience: str,
    required_scopes: list[str] | None,
    expected_zone_id: str | None,
    revocations: RevocationStore,
    require_agent: bool = False,
    require_delegation: bool = False,
    require_chain_contains: list[str] | None = None,
    max_hop_count: int | None = None,
    required_targets: list[str] | None = None,
    required_use: str | None = "per_call",
) -> AuthResult:
    if not token:
        return AuthResult(None, AuthError("missing_token", "Missing bearer token"))

    cfg = JwtConfig(
        issuer=issuer,
        audience=audience,
        expected_zone_id=expected_zone_id,
        required_scopes=required_scopes or [],
        required_targets=required_targets or [],
        required_use=required_use,
        require_agent=require_agent,
        require_delegation=require_delegation,
        require_chain_contains=require_chain_contains or [],
        max_hop_count=max_hop_count,
    )

    try:
        claims = await verify_config(token, cfg)
    except ScopeInsufficientError as err:
        return AuthResult(None, AuthError("insufficient_scope", str(err)))
    except AgentIdentityRequiredError:
        return AuthResult(None, AuthError("agent_required", "Agent identity required"))
    except DelegationRequiredError:
        return AuthResult(None, AuthError("delegation_required", "Delegation required"))
    except ChainMismatchError as err:
        return AuthResult(None, AuthError("chain_mismatch", str(err)))
    except HopCountExceededError as err:
        return AuthResult(None, AuthError("hop_count_exceeded", str(err)))
    except ZoneInvalidError:
        return AuthResult(None, AuthError("invalid_zone", "Token zone validation failed"))
    except TokenInvalidError:
        return AuthResult(None, AuthError("invalid_token", "Token validation failed"))

    if revocations is None:
        return AuthResult(None, AuthError("invalid_token", "Revocation store required"))
    active_error = check_active_authority(claims, revocations)
    if active_error is not None:
        return AuthResult(None, active_error)

    return AuthResult(claims, None)


def check_active_authority(claims: object, revocations: RevocationStore, now_seconds: int | None = None) -> AuthError | None:
    import time

    sid = getattr(claims, "sid", "")
    if not sid:
        return AuthError("invalid_token", "Token validation failed")
    expires_at = getattr(claims, "expires_at", 0)
    if expires_at and expires_at <= (now_seconds if now_seconds is not None else int(time.time())):
        return AuthError("invalid_token", "Token expired during execution")
    for anchor in _revocation_anchors(claims):
        if revocations.is_revoked(anchor):
            return AuthError("session_revoked", "Session revoked")
    return None


def _revocation_anchors(claims: object) -> list[str]:
    anchors = [
        getattr(claims, "sid", None),
        getattr(claims, "root_sid", None),
        getattr(claims, "agent_session_id", None),
        getattr(claims, "delegation_edge_id", None),
    ]
    out: list[str] = []
    for anchor in anchors:
        if isinstance(anchor, str) and anchor and anchor not in out:
            out.append(anchor)
    return out

File: packages/verify/ts/src/authenticate.ts
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// Framework-neutral mandate verification: bearer verify, revocation check, typed result.

import {
  AgentIdentityRequiredError,
  ChainMismatchError,
  DelegationRequiredError,
  HopCountExceededError,
  ScopeInsufficientError,
  TokenInvalidError,
  ZoneInvalidError,
  MANDATE_USE_RESOURCE,
  verify,
  warmJwks,
  type JwtConfig,
  type JwksCache,
} from '@caracalai/identity'
import type { RevocationStore } from '@caracalai/revocation'
import type { AuthError, AuthErrorCode, AuthResult, Principal } from './types.js'

export type AuthDeps = JwtConfig & { revocations: RevocationStore; jwksCache?: JwksCache }
export type AuthOverrides = Partial<Omit<AuthDeps, 'issuer' | 'audience' | 'revocations' | 'jwksCache'>>

export interface MandateVerifier {
  readonly defaults: AuthDeps
  authenticate: (token: string, overrides?: AuthOverrides) => Promise<AuthResult>
  authorization: (authHeader: string | undefined, overrides?: AuthOverrides) => Promise<AuthResult>
  require: (overrides: AuthOverrides) => MandateVerifier
  warmup: () => Promise<void>
}

const BEARER_SCHEME = 'bearer'

export function extractBearer(authHeader: string | undefined): string | null {
  if (authHeader === undefined || authHeader.slice(0, BEARER_SCHEME.length).toLowerCase() !== BEARER_SCHEME) return null
  const value = authHeader.slice(BEARER_SCHEME.length)
  if (value.length === value.trimStart().length) return null
  const token = value.trim()
  return token === '' ? null : token
}

export function createMandateVerifier(defaults: AuthDeps): MandateVerifier {
  return {
    defaults,
    authenticate(token: string, overrides: AuthOverrides = {}): Promise<AuthResult> {
      return authenticate(token, { ...defaults, ...overrides })
    },
    authorization(authHeader: string | undefined, overrides: AuthOverrides = {}): Promise<AuthResult> {
      const token = extractBearer(authHeader)
      return authenticate(token ?? '', { ...defaults, ...overrides })
    },
    require(overrides: AuthOverrides): MandateVerifier {
      return createMandateVerifier({ ...defaults, ...overrides })
    },
    async warmup(): Promise<void> {
      // JWKS keysets are zone-scoped; without a configured zone the keyset to
      // warm is unknown until the first token arrives.
      if (!defaults.zoneId) return
      if (defaults.jwksCache) {
        await defaults.jwksCache.warm(defaults.issuer, defaults.zoneId)
        return
      }
      await warmJwks(defaults.issuer, defaults.zoneId)
    },
  }
}

export async function authenticate(token: string, deps: AuthDeps): Promise<AuthResult> {
  if (!token) {
    return { ok: false, error: authError('missing_token') }
  }

  try {
    const { revocations, ...jwtConfig } = deps
    const claims = await verify(token, { ...jwtConfig, requiredUse: jwtConfig.requiredUse ?? MANDATE_USE_RESOURCE })
    if (!revocations || typeof revocations.isRevoked !== 'function') {
      return { ok: false, error: authError('invalid_token', 'Revocation store required') }
    }
    const activeError = await checkActiveAuthority(claims, revocations)
    if (activeError) {
      return { ok: false, error: activeError }
    }
    return { ok: true, principal: claims }
  } catch (err) {
    if (err instanceof ScopeInsufficientError) {
      return { ok: false, error: authError('insufficient_scope', err.message) }
    }
    if (err instanceof AgentIdentityRequiredError) {
      return { ok: false, error: authError('agent_required', err.message) }
    }
    if (err instanceof DelegationRequiredError) {
      return { ok: false, error: authError('delegation_required', err.message) }
    }
    if (err instanceof ChainMismatchError) {
      return { ok: false, error: authError('chain_mismatch', err.message) }
    }
    if (err instanceof HopCountExceededError) {
      return { ok: false, error: authError('hop_count_exceeded', err.message) }
    }
    if (err instanceof ZoneInvalidError) {
      return { ok: false, error: authError('invalid_zone') }
    }
    if (err instanceof TokenInvalidError) {
      return { ok: false, error: authError('invalid_token') }
    }
    return { ok: false, error: authError('invalid_token') }
  }
}

export async function checkActiveAuthority(claims: Principal, revocations: RevocationStore, nowMs = Date.now()): Promise<AuthError | null> {
  if (!claims.sid) {
    return authError('invalid_token')
  }
  if (claims.expiresAt * 1000 <= nowMs) {
    return authError('invalid_token', 'Token expired during execution')
  }
  const checks = await Promise.all(revocationAnchors(claims).map((anchor) => revocations.isRevoked(anchor)))
  if (checks.some(Boolean)) return authError('session_revoked')
  const epochError = await graphEpochError(claims, revocations)
  if (epochError) return epochError
  return null
}

function revocationAnchors(claims: Principal): string[] {
  const anchors = [claims.sid, claims.rootSid, claims.agentSessionId, claims.delegationEdgeId].filter(
    (value): value is string => typeof value === 'string' && value !== '',
  )
  return [...new Set(anchors)]
}

async function graphEpochError(claims: Principal, revocations: RevocationStore): Promise<AuthError | null> {
  if (claims.graphEpoch === undefined || !revocations.currentDelegationEpoch) return null
  const currentEpoch = await revocations.currentDelegationEpoch(claims.zoneId)
  return currentEpoch > claims.graphEpoch ? authError('delegation_stale') : null
}

function authError(code: AuthError['code'], description = defaultDescription(code)): AuthError {
  return { code, description, hint: defaultHint(code) }
}

/**
 * Canonical HTTP status for an authentication failure code. This is the single
 * source of truth for every HTTP adapter so the boundary semantics stay
 * identical across frameworks and languages.
 *
 * 401 means the credential itself was not accepted (missing, malformed, wrong
 * zone, revoked, or stale). 403 means the mandate verified but the authority it
 * carries is insufficient for the route (missing scope, wrong principal kind, or
 * an unmet delegation requirement).
 */
export function httpStatusForAuthError(code: AuthErrorCode): number {
  switch (code) {
    case 'insufficient_scope':
    case 'agent_required':
    case 'delegation_required':
    case 'chain_mismatch':
    case 'hop_count_exceeded':
      return 403
    default:
      return 401
  }
}

function defaultDescription(code: AuthError['code']): string {
  switch (code) {
    case 'missing_token':
      return 'Missing bearer token'
    case 'invalid_zone':
      return 'Token zone validation failed'
    case 'insufficient_scope':
      return 'Required scope is missing'
    case 'session_revoked':
      return 'Session revoked'
    case 'delegation_stale':
      return 'Delegation graph changed'
    case 'agent_required':
      return 'Agent identity required'
    case 'delegation_required':
      return 'Delegation required'
    case 'chain_mismatch':
      return 'Delegation chain validation failed'
    case 'hop_count_exceeded':
      return 'Hop count exceeded'
    default:
      return 'Token validation failed'
  }
}

function defaultHint(code: AuthError['code']): string {
  switch (code) {
    case 'missing_token':
      return 'Send Authorization: Bearer <Caracal mandate>.'
    case 'invalid_zone':
      return 'Check the configured zoneId and the mandate zone_id claim.'
    case 'insufficient_scope':
      return 'Request a mandate that includes every required scope for this route.'
    case 'session_revoked':
      return 'Refresh the mandate or start a new authorized session.'
    case 'delegation_stale':
      return 'Refresh the mandate so delegated authority is evaluated against the latest graph.'
    case 'agent_required':
      return 'Use an agent-issued resource mandate for this endpoint.'
    case 'delegation_required':
      return 'Use a mandate produced by a delegated grant flow.'
    case 'chain_mismatch':
      return 'Check requireChainContains and the mandate delegation chain.'
    case 'hop_count_exceeded':
      return 'Reduce delegation depth or raise maxHopCount deliberately.'
    default:
      return 'Check issuer, audience, signature, expiry, token use, scopes, and targets.'
  }
}
